src/wasm/wasm-type.cpp

```diff
@@ -155,18 +163,57 @@ struct TypePrinter {
   std::ostream& print(const Rtt& rtt);
 
 private:
-  template<typename T, typename F> std::ostream& printChild(T curr, F printer) {
-    auto it = depths.find(curr.getID());
-    if (it != depths.end()) {
-      assert(it->second <= currDepth);
-      size_t relativeDepth = currDepth - it->second;
-      return os << "..." << relativeDepth;
-    }
-    depths[curr.getID()] = ++currDepth;
-    printer();
-    depths.erase(curr.getID());
-    return os;
-  }
+  template<typename T, typename F> std::ostream& printChild(T curr, F printer);
+};
+
+// Helper for hashing the shapes of TypeInfos and HeapTypeInfos. Keeps track of
+// previously seen HeapTypes to avoid traversing them more than once. Infos
+// referring to different type IDs but sharing a finite shape will compare and
+// hash the same.
+struct FiniteShapeHasher {
+  bool topLevelOnly;
+  size_t currDepth = 0;
+  size_t currStep = 0;
+  std::unordered_map<HeapType, size_t> seen;
+
+  FiniteShapeHasher(bool topLevelOnly = false) : topLevelOnly(topLevelOnly) {}
+
+  size_t hash(Type type);
+  size_t hash(HeapType heapType);
+  size_t hash(const TypeInfo& info);
+  size_t hash(const HeapTypeInfo& info);
+  size_t hash(const Tuple& tuple);
+  size_t hash(const Field& field);
+  size_t hash(const Signature& sig);
+  size_t hash(const Struct& struct_);
+  size_t hash(const Array& array);
+  size_t hash(const Rtt& rtt);
+};
+
+// Helper for comparing the shapes of TypeInfos and HeapTypeInfos for equality.
+// Like FiniteShapeHasher, keeps track of previously seen HeapTypes. Note that
+// this does not test for coinductive equality of the infinite expansion of the
+// type tree, but rather tests for equality of the finite shape of the graph. If
+// FiniteShapeEquator reports that two type shapes are equal, FiniteShapeHasher
+// should produce the same hash for them.
+struct FiniteShapeEquator {
```

Comment:
We have `Comparer` and `Comparator` in other places, but not sure those are better than `Equator`. But it would be good to standardize on the best at some point.

Reply:
Yeah, makes sense. In this file we also have a `TypeComparator`, but that implements `<` rather than `==`, so I thought a different name would be good for making that difference clear.